uqcsbot/remindme.py

```diff
@@ -0,0 +1,408 @@
+import datetime as dt
+import discord
+from discord import app_commands
+from discord.ext import commands
+from functools import partial
+import logging
+import random
+from typing import List, NamedTuple, Optional, Union
+
+from uqcsbot.bot import UQCSBot
+from uqcsbot.models import Reminders
+
+
+USER_REMINDER_LIMIT = 10
+REMINDER_IDS_RANGE = (1, 100)  # range from which reminder ids will be randomly chosen
+
+REMINDME_MESSAGE_TITLE = "RemindMe: Reminder"
+REMINDME_ERROR_TITLE = "RemindMe: Error"
+REMINDME_ADDED_TITLE = "RemindMe: Reminder Added"
+REMINDME_REMOVED_TITLE = "RemindMe: Reminder Removed"
+REMINDME_LIST_TITLE = "RemindMe: List Reminders"
+
+REMINDER_MESSAGE = "Reminder set by <@{}>:\n> {}"
+REMINDER_NOT_FOUND_ERR = "Reminder id not found."
+REMINDER_LIMIT_REACHED_ERR = f"You've reached the maximum number of active reminders ({USER_REMINDER_LIMIT})."
+DATETIME_VALID_FORMAT_ERR = "Dates and times should conform to a valid ISO 8601 format."
+DATETIME_IN_PAST_ERR = "Datetime can't be in the past."
+STARTEND_DATE_IN_PAST_ERR = "Start or end date can't be in the past."
+END_DATE_BEFORE_START_ERR = "End date can't be before the start date."
+NEGATIVE_WEEK_FREQUENCY_ERR = "Week frequency must be non-negative."
+
+REMOVE_REMINDERS_FOOTER = "You can remove active reminders with /remindme remove."
+LIST_REMINDERS_FOOTER = "You can view your active reminders with /remindme list."
+
+DISPLAY_DATE_FORMAT = '%A %d %b %Y'
+DISPLAY_DAY_NAME_FORMAT = '%A'
+DISPLAY_TIME_FORMAT = '%-I:%M %p'
+
+
+class Reminder(NamedTuple):
+    id: int
+    user_id: int
+    channel_id: int | None  # None => send reminder in DMs
+    time_created: dt.datetime
+    message: str
+    time: dt.time
+    start_date: dt.date
+    end_date: dt.date | None  # None => non-ending
+    week_frequency: int | None  # 0 => daily, None => one-time
+
+    def __str__(self):
+        id = self.id
```

Comment:
Use of reminder id here - outputting reminders publicly will display a lot of duplicate IDs, which isn't a problem internally but may confuse people

Reply:
That's true, it would probably be clearer for people if I made ids unique across all reminders rather than just for a specific user. I was also thinking of implementing admin commands at some point (so admins can remove others' reminders), so having unique ids across the board would make that easier as well.